internal/os/disk/api.go

```diff
@@ -294,20 +248,39 @@ func (imp DiskAPI) GetDiskNumberWithID(page83ID string) (uint32, error) {
 	json.Unmarshal([]byte(outString), &disks)
 
 	for i := range disks {
-		h, err := syscall.Open(disks[i].Path, syscall.O_RDONLY, 0)
+		diskNumber, diskPage83ID, err := imp.GetDiskNumberAndPage83ID(disks[i].Path)
 		if err != nil {
 			return 0, err
 		}
 
-		found, err := imp.DiskHasPage83ID(h, page83ID)
-		if found {
-			return imp.GetDiskNumber(h)
+		if diskPage83ID == page83ID {
+			return diskNumber, nil
 		}
 	}
 
 	return 0, fmt.Errorf("Could not find disk with Page83 ID %s", page83ID)
 }
 
+func (imp DiskAPI) GetDiskNumberAndPage83ID(path string) (uint32, string, error) {
+	h, err := syscall.Open(path, syscall.O_RDONLY, 0)
+	defer syscall.Close(h)
+	if err != nil {
+		return 0, "", err
+	}
+
+	diskNumber, err := imp.GetDiskNumber(h)
+	if err != nil {
+		return 0, "", err
```

Comment:
A DiskNumber of `0` is a valid ID for a disk enumerated as \\.\PhysicalDrive0. So returning 0 for the error scenario is not recommended. Maybe return a pointer to uint32 or -1?

Reply:
hmm I would expect caller to always check the error first. -1 i can't do since it's unsigned (or caller needs to convert to unsigned), pointer maybe I could do but then the caller needs to always check if it's nil before dereference, and I would argue in that case they should just check the error

Reply:
Looks like in other functions we return 0 in the error case too. So makes sense to check error first.